`Development`: Add Helios push based lifecycle monitoring with manual DB migration status events

egekocabas opened the pull request:
### Checklist
#### General
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).


### Motivation and Context
> **Note:** Before diving into this PR’s details, please take a moment to read the library’s [README](https://github.com/ls1intum/Helios/blob/staging/server/helios-status-spring-starter/PACKAGE_README.md)


Helios currently **pulls** each environment’s status; this library flips that to a **push** model:
- Services actively send their lifecycle state to one or more Helios endpoints.
- We cover `STARTING_UP`, `RUNNING` heart-beats, `SHUTTING_DOWN`, and `FAILED`.
- **By default the starter does NOT watch your database migrations**—this PR adds explicit, manual pushes for:
  - `DB_MIGRATION_STARTED`
  - `DB_MIGRATION_FINISHED`
  - `DB_MIGRATION_FAILED`

### Description
- Added dependency `de.tum.cit.aet:helios-status-spring-starter` (~22 KB)
- New `helios.status.*` YAML block (see example below).
- **Runtime behaviour**
  - `BootLifecycleListener` inside the starter pushes Spring lifecycle events; `HeartbeatScheduler` sends `RUNNING` every 30s.

- **DB migration integration**
  - Since Liquibase don’t expose events, we inject a  `Optional<HeliosClient>` into `DatabaseMigration` and `LiquibaseConfiguration` to call:
    ```java
    helios.pushDbMigrationStarted();
    // … run migration …
    helios.pushDbMigrationFinished();
    // on exception:
    helios.pushDbMigrationFailed();
    ```

- Example `application.yml` (The YAML is set and populated to the nodes in the deployment)
```yaml
helios:
  status:
    enabled: false         # ← master switch. false = do nothing (default)
    environment-name: ""   # ← MUST equal the exact GitHub Actions environment name
    endpoints:             # ← one or more Helios instances to push to
      - url: https://helios.aet.cit.tum.de/api/environments/status            #     instance REST URL (…/api/environments/status)
        secret-key: ${HELIOS_PROD_SECRET_KEY}     #     repo-specific secret; generated in Helios UI
      # - url: …           #   add a second entry for prod, etc.
```


key | required? | purpose & rules
-- | -- | --
enabled | no (default false) | Turns the feature on. Keep it false everywhere until the repo has a secret and the environment in Helios is configured for Push Update.
environment-name | yes when enabled | Must match the exact name of the GitHub environment ( not the display label shown in Helios ).Examples: artemis-test1.artemis.cit.tum.de, artemis.tum.de, …
endpoints | yes when enabled | At least one url = Helios REST endpoint, ending in `/api/environments/status`. secret-key = repo-specific token generated in the Helios UI. Add a second entry if you also want to push to staging, etc.


### Steps for Testing
#### Prerequisites:
- 1 user with WRITE access to the Artemis GitHub repository 

#### Steps
- Open https://helios.aet.cit.tum.de/ and select Artemis.
- Find this PR and deploy to one of the test servers
- See the status badge changes while deploying (`Shutting down`, `Migrating DB`, `Migration finished`, `Running`)
- Open the test server and validate it is up and functional
- (Optional): See the test server logs and search for `helios`

### Screenshots
- A few points to note:
  - The environment list view in Helios is updated at regular intervals. This means that if multiple status updates are pushed within a short time frame, only the most recent one may be visible. For example, during application startup, the Starting up event might be immediately followed by Migrating DB, and only the latter will appear in the UI.
  - Additionally, the manual dispatch of database-related events may occur too close to the application's startup lifecycle event. Since the library hooks into Spring lifecycle events, it's possible that both events are sent almost simultaneously, causing the Starting up status to be overwritten before it can be displayed.

https://github.com/user-attachments/assets/c0214a6c-a236-43ad-b748-483aae5762e0


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

- **New Features**
  - Added integration with Helios for push-based database migration status updates, providing real-time notifications on migration progress and failures.

- **Chores**
  - Updated dependencies and configuration to support Helios integration.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

egekocabas commented on src/main/resources/config/application.yml:
```diff
@@ -366,3 +366,14 @@ eureka:
+# Helios configuration
+helios:
+  status:
+    enabled: true
```
If we would change it to false, then please propose where to enable it

b-fein replied:
In your actual production deployment configuration -> your Ansible setup.

Hialus reviewed (changes requested):
Currently, this PR seems to require mandatory configuration changes for every (Docker-based) Artemis instance. If the Helios config is not set at all, the Artemis server does not start.
```
artemis-app-node-1  | Error starting ApplicationContext. To display the condition evaluation report re-run your application with 'debug' enabled.
artemis-app-node-1  | 2025-07-01T14:36:58.586Z  WARN 1 --- [Artemis] [           main] de.tum.cit.aet.helios.HeliosClient       : Helios payload is missing environment. Skipping push.
artemis-app-node-1  | 2025-07-01T14:36:58.614Z ERROR 1 --- [Artemis] [           main] o.s.boot.SpringApplication               : Application run failed
artemis-app-node-1  |
artemis-app-node-1  | java.lang.IllegalStateException: helios.status.enabled=true but environment-name or endpoints missing.
artemis-app-node-1  | Each endpoint needs url + secretKey.
artemis-app-node-1  | 	at de.tum.cit.aet.helios.autoconfig.HeliosStatusAutoConfiguration.lambda$validateHeliosCfg$0(HeliosStatusAutoConfiguration.java:63)
artemis-app-node-1  | 	at org.springframework.boot.SpringApplication.lambda$callRunner$4(SpringApplication.java:784)
artemis-app-node-1  | 	at org.springframework.util.function.ThrowingConsumer$1.acceptWithException(ThrowingConsumer.java:82)
artemis-app-node-1  | 	at org.springframework.util.function.ThrowingConsumer.accept(ThrowingConsumer.java:60)
artemis-app-node-1  | 	at org.springframework.util.function.ThrowingConsumer$1.accept(ThrowingConsumer.java:86)
artemis-app-node-1  | 	at org.springframework.boot.SpringApplication.callRunner(SpringApplication.java:796)
artemis-app-node-1  | 	at org.springframework.boot.SpringApplication.callRunner(SpringApplication.java:784)
artemis-app-node-1  | 	at org.springframework.boot.SpringApplication.lambda$callRunners$3(SpringApplication.java:772)
artemis-app-node-1  | 	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
artemis-app-node-1  | 	at java.base/java.util.stream.SortedOps$SizedRefSortingSink.end(SortedOps.java:357)
artemis-app-node-1  | 	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:510)
artemis-app-node-1  | 	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
artemis-app-node-1  | 	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
artemis-app-node-1  | 	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
artemis-app-node-1  | 	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
artemis-app-node-1  | 	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
artemis-app-node-1  | 	at org.springframework.boot.SpringApplication.callRunners(SpringApplication.java:772)
artemis-app-node-1  | 	at org.springframework.boot.SpringApplication.run(SpringApplication.java:325)
artemis-app-node-1  | 	at de.tum.cit.aet.artemis.ArtemisApp.main(ArtemisApp.java:81)
artemis-app-node-1  | 	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
artemis-app-node-1  | 	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
artemis-app-node-1  | 	at org.springframework.boot.loader.launch.Launcher.launch(Launcher.java:102)
artemis-app-node-1  | 	at org.springframework.boot.loader.launch.Launcher.launch(Launcher.java:64)
artemis-app-node-1  | 	at org.springframework.boot.loader.launch.WarLauncher.main(WarLauncher.java:53)
artemis-app-node-1  |
artemis-app-node-1  | 2025-07-01T14:36:58.637Z  WARN 1 --- [Artemis] [           main] de.tum.cit.aet.helios.HeliosClient       : Helios payload is missing environment. Skipping push.
artemis-app-node-1  | 2025-07-01T14:37:00.787Z  INFO 1 --- [Artemis] [           main] de.tum.cit.aet.helios.HeliosClient       : Destroying HeliosClient, flushing pending pushes.
artemis-app-node-1  | 2025-07-01T14:37:00.829Z  INFO 1 --- [Artemis] [           main] d.t.c.a.a.c.config.CacheConfiguration    : Closing Cache Manager
artemis-app-node-1  | 2025-07-01T14:37:03.967Z  INFO 1 --- [Artemis] [os-ShutdownHook] de.tum.cit.aet.helios.HeliosClient       : Helios shutdown-hook: flushing pending pushes
```

This is a breaking configuration change! If the Helios config is not manually specified, it should instead default to Helios status updates being disabled. No one outside of TUM is using Helios for deployment, so they also should not be forced to manually disable Helios in the config. 
In short: The Helios status updating should be opt-in in the config, not opt-out.

eylulnc reviewed (approved):
Tested by deploying TS2 and TS4
<img width="1364" alt="Screenshot 2025-07-03 at 18 43 27" src="https://github.com/user-attachments/assets/bbdcd101-b25c-40bf-bec5-16693db655d0" />
<img width="1364" alt="Screenshot 2025-07-03 at 18 43 32" src="https://github.com/user-attachments/assets/0bd4d6ce-474b-4f45-9d85-1eed6d192df5" />

Different statuses are observed

konrad2002 reviewed (approved):
tested on TS3, works as expected

TurkerKoc reviewed (approved):
Deployed and tested on test-server 5. Everything works as expected with status updates for db migrations. 💯

chuuuun reviewed (approved):
Tested on ts6. Worked both in described situation and restart the server on the half way of deployment.

badkeyy reviewed (approved):
Tested by deploying to TS1. Helios tags update as described. Code lgtm.

egekocabas commented:
Thank you everybody for testing the PR 🙏🏻 

> Code changes in Artemis are looking very good 👍 I have just noticed that the Helios Status Client depends on `okhttp3` which is another external framework. As we like to avoid adding too many different dependencies (which are hard to maintain from a security point of view), I would highly prefer when the Helios Status Client could use Spring's `RestClient` which is integrated into Spring anyway.

I have refactored the library and removed the okhttp3 dependency. Now we are using spring boot's `RestClient`. See the https://github.com/ls1intum/Helios/pull/803 PR for the changes 👍🏻 (Also created a changelog file to keep track of the library changes [CHANGELOG.md](https://github.com/ls1intum/Helios/blob/staging/server/helios-status-spring-starter/CHANGELOG.md))

Tested on TS 1,2,3,4,5,6 👍🏻

eylulnc reviewed (approved):
Reapprove manual testing, mentioned states are observed.
<img width="1366" alt="Screenshot 2025-07-08 at 01 14 37" src="https://github.com/user-attachments/assets/09534826-f1d3-4787-8248-95c8d6a3c843" />
